### PR TITLE
New config object base and apiVersion

### PR DESCRIPTION
**Pull request by paulasjes-stripe**
Part 1 of https://github.com/stripe/stripe-node/pull/703

This adds the ability to pass in a configuration object as the 2nd parameter of `Stripe`. For example:
```js
const stripe = new Stripe('sk_test_123', {
  apiVersion: '2019-09-09',
)};
```

This allows the old behavior to still work:
```js
const stripe = new Stripe('sk_test_123', '2019-09-09');
```

Additionally this behavior will also still work:
```js
const stripe = new Stripe('sk_test_123');
stripe.setApiVersion('2019-09-09');
```

With the idea being of eventually deprecating `stripe.setApiVersion` in favor of the config object.

r? @rattrayalex-stripe @ob-stripe @richardm-stripe 
cc @stripe/api-libraries

### CODE REVIEW

File: lib/stripe.js [hunk 1]
```diff
@@ -251,6 +261,50 @@ Stripe.prototype = {
...
   },
+
+  _getPropsFromConfig(config) {
+    // config can be an object or a string
+    const isString = typeof config === 'string';
+    const isObject = config === Object(config) && !Array.isArray(config);
+
+    if (!isObject && !isString) {
+      throw new Error('Config must either be an object or a string');
+    }
```
Comment (richardm-stripe) [inline lib/stripe.js]: nit: What do you think about checking *all four* cases (undefined, string, object, other) in a single place, and producing the canonical form {version: string} immediately, then performing the version string validation along with the other validation?

File: lib/stripe.js [hunk 2]
```diff
@@ -251,6 +261,50 @@ Stripe.prototype = {
       this[utils.pascalToCamelCase(name)] = new resources[name](this);
     }
   },
+
+  _getPropsFromConfig(config) {
+    // config can be an object or a string
+    const isString = typeof config === 'string';
+    const isObject = config === Object(config) && !Array.isArray(config);
+
+    if (!isObject && !isString) {
```
Comment (rattrayalex-stripe) [inline lib/stripe.js]: part of me thinks we should accept `null` as valid, and treat it as the same as `undefined` (that is, don't set `apiVersion`)

File: lib/stripe.js [hunk 3]
```diff
@@ -251,6 +261,50 @@ Stripe.prototype = {
...
+      throw new Error(
+        'Api version string should be in the format YYYY-MM-DD or "latest"'
+      );
+    }
+
+    let props = {};
+
+    // If config is a string, we assume the old behavior of passing in a string representation of the api version
+    if (isString) {
+      props.apiVersion = config;
```
Comment (rattrayalex-stripe) [inline lib/stripe.js]: nit, I think it might make sense to early-return `{apiVersion: config}` here, since it'll be impossible to set other properties later anyway. 

Then, later on, we can just `return config`

File: lib/utils.js [hunk 4]
```diff
@@ -356,6 +356,10 @@ const utils = (module.exports = {
       return v.toString(16);
     });
   },
+
+  // Only allow versions in the format YYYY-MM-DD
```
[the remark this answers is not in the document]
Comment (paulasjes-stripe) [reply lib/utils.js]: TIL! Is it conceivable that api versions could take the form of `foo=bar YYYY-MM-DD` or would we only expect additional parameters after the date string?

File: lib/stripe.js (hunk 1 above)
Comment (paulasjes-stripe) [reply lib/stripe.js]: Damnit that's so much better

File: lib/utils.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (paulasjes-stripe) [reply lib/utils.js]: Also is something like this possible: `YYYY-MM-DD foo_beta=v1 bar_beta=v2`?

I'm thinking that maybe the regex just checks for the date at the beginning of the string and doesn't care about what comes after.

File: lib/stripe.js (hunk 1 above)
Comment (paulasjes-stripe) [reply lib/stripe.js]: I rejigged a bit and think this "return fast and early" strategy works, let me know what you think.

**Comment by paulasjes-stripe** [comment]
Addressed your feedback and added more tests for edge cases, ptal @rattrayalex-stripe @richardm-stripe

File: lib/utils.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (rattrayalex-stripe) [reply lib/utils.js]: I think I'd rather forego regex checks unless they're really necessary, and rely on the api to tell the user that the string doesn't fit. Just buys us the most flexibility. Thoughts?

File: lib/stripe.js [hunk 5]
```diff
@@ -251,6 +261,53 @@ Stripe.prototype = {
...
+
+    // If config is a string, we assume the old behavior of passing in a string representation of the api version
+    if (isString) {
+      return {
+        apiVersion: config,
+      };
+    }
+
+    // If config is an object, we assume the new behavior and make sure it doesn't contain any unexpected values
+    if (isObject) {
```
Comment (rattrayalex-stripe) [inline lib/stripe.js]: nit, this conditional is unnecessary (fine to leave though)

File: lib/utils.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (paulasjes-stripe) [reply lib/utils.js]: I like the concept of failing as soon as possible if something is obviously wrong. If I were to initialize stripe-node like so:

```js
const stripe = new Stripe('sk_test_123', 'bananas');
```

Then stripe-node would happily return a version of Stripe with a bogus api version. The error won't occur until we actually attempt to make a request.

I think that this sort of clearly wrong input should be caught and handled before any request is made. 

Saying that, if we have too many API version format options and edge cases, then I'm happy to let the API take care of that for us.

[the remark this answers is not in the document]
Comment (rattrayalex-stripe) [reply lib/utils.js]: Ah sorry, missed this comment earlier – earlier error messages are certainly better than later ones, but the error message from the API about an invalid version string should come pretty quick, no? 

Furthermore the client can't have enough information to make an accurate decision – we can't disallow a valid date which isn't a valid api version, and may indeed allow `"bananas"` at one point in the future for a very special beta program!

**Comment by paulasjes-stripe** [comment]
@rattrayalex-stripe You've convinced me that the regex is a bad idea, so I removed it. Have a look now

**Comment by richardm-stripe** [comment]
lgtm
